pkg/inventory/model/journal.go

```diff
@@ -0,0 +1,293 @@
+package model
+
+import (
+	liberr "github.com/konveyor/controller/pkg/error"
+	"github.com/konveyor/controller/pkg/ref"
+	"reflect"
+	"sync"
+)
+
+//
+// Event Actions.
+var (
+	Created int8 = 0x01
+	Updated int8 = 0x02
+	Deleted int8 = 0x04
+)
+
+//
+// Model event.
+type Event struct {
+	// The event subject.
+	Model Model
+	// The event action (created|updated|deleted).
+	Action int8
+}
+
+//
+// Event handler.
+type EventHandler interface {
+	// A model has been created.
+	Created(Model)
+	// A model has been updated.
+	Updated(Model)
+	// A model has been deleted.
+	Deleted(Model)
+	// An error has occurred delivering an event.
+	Error(error)
+	// An event watch has ended.
+	End()
+}
+
+//
+// Model event watch.
+type Watch struct {
+	// Model to be watched.
+	Model Model
+	// Event handler.
+	Handler EventHandler
+	// Event queue.
+	queue chan *Event
+	// Started
+	started bool
+}
+
+//
+// Match by model `kind`.
+func (w *Watch) Match(model Model) bool {
+	return ref.ToKind(w.Model) == ref.ToKind(model)
+}
+
+//
+// Queue event.
+func (w *Watch) notify(event *Event) {
+	if !w.Match(event.Model) {
+		return
+	}
+	defer func() {
+		recover()
+	}()
+	select {
+	case w.queue <- event:
+	default:
+		err := liberr.New("full queue, event discarded")
+		w.Handler.Error(err)
+	}
+}
+
+//
+// Run the watch.
+// Forward events to the `handler`.
+func (w *Watch) Start() {
+	if w.started {
+		return
+	}
+	run := func() {
+		for event := range w.queue {
+			switch event.Action {
+			case Created:
+				w.Handler.Created(event.Model)
+			case Updated:
+				w.Handler.Updated(event.Model)
+			case Deleted:
+				w.Handler.Deleted(event.Model)
+			default:
+				w.Handler.Error(liberr.New("unknown action"))
+			}
+		}
+		w.Handler.End()
+	}
+
+	w.started = true
+	go run()
+}
+
+//
+// End the watch.
+func (w *Watch) End() {
+	close(w.queue)
+}
+
+//
+// Event manager.
+type Journal struct {
+	*Client
+	mutex sync.RWMutex
+	// List of registered watches.
+	watch []*Watch
+	// Queue of staged events.
+	staged []*Event
+	// Enabled.
+	enabled bool
+}
+
+//
+// The journal is enabled.
+// Must be enabled for watch models.
+func (r *Journal) Enabled() bool {
+	r.mutex.RLock()
+	defer r.mutex.RUnlock()
+	return r.enabled
+}
+
+//
+// The journal is enabled.
+// Must be enabled for watch models.
+// Disable will end all watches and discard staged events.
+func (r *Journal) Enable(enabled bool) {
```

Comment:
This function feels a little weird to me, since it's called `Enable` but it also disables, and at a glance it's easy to confuse the parameter `enabled` with the instance variable `r.enabled`. The multiple levels of control flow also make it a little tough to read. What would you think about splitting it into `Enable()` and `Disable()`? Something like this:

```go
func (r *Journal) Enable() {
   r.mutex.Lock()
   defer r.mutex.Unlock()

   if r.enabled {
      return
   }

   r.staged = []*Event{}
   r.enabled = true
}

func (r *Journal) Disable() {
   r.mutex.Lock()
   defer r.mutex.Unlock()

   if !r.enabled {
      return
   }

   for _, w := range r.watch {
      w.End()
   }
   r.watch = []*Watch{}
   r.staged = []*Event{}
   r.enabled = false
}
```

Reply:
Agreed.  What you suggested is clearer and will update the PR.